src/keys/slip10.rs

```diff
@@ -7,79 +7,168 @@ use alloc::vec::Vec;
 use core::convert::TryFrom;
 
 use serde::{Deserialize, Serialize};
-use zeroize::{Zeroize, ZeroizeOnDrop, Zeroizing};
+use zeroize::{Zeroize, ZeroizeOnDrop};
 
 use crate::macs::hmac::HMAC_SHA512;
-#[cfg(feature = "ed25519")]
-use crate::signatures::ed25519;
-#[cfg(feature = "secp256k1")]
-use crate::signatures::secp256k1_ecdsa;
 
 // https://github.com/satoshilabs/slips/blob/master/slip-0010.md
 // https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
 // https://en.bitcoin.it/wiki/BIP_0039
 
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
-pub enum Curve {
-    #[cfg(feature = "ed25519")]
-    Ed25519,
-    #[cfg(feature = "secp256k1")]
-    Secp256k1,
+pub mod hazmat {
+    pub trait Derivable {
+        fn is_key_valid(key_bytes: &[u8; 33]) -> bool;
+        fn into_key(key_bytes: &[u8; 33]) -> Self;
+        fn add_key(key_bytes: &mut [u8; 33], parent_key: &[u8; 33]) -> bool;
+
+        const ALLOW_NON_HARDENED: bool;
+        fn calc_non_hardened_data(key_bytes: &[u8; 33]) -> [u8; 33];
+    }
+    pub trait IsSecretKey: Derivable {
+        const SEEDKEY: &'static [u8];
+        // PublicKey type may not be Derivable as is the case with ed25519
+        type PublicKey;
+    }
+    pub trait IsPublicKey: Derivable {
+        type SecretKey: IsSecretKey;
+    }
 }
 
-impl Curve {
-    pub fn is_non_hardened_supported(&self) -> bool {
-        match self {
-            #[cfg(feature = "ed25519")]
-            Curve::Ed25519 => false,
-            #[cfg(feature = "secp256k1")]
-            Curve::Secp256k1 => true,
+pub use hazmat::{Derivable, IsPublicKey, IsSecretKey};
+
+#[cfg(feature = "ed25519")]
+pub mod ed25519 {
+    use super::hazmat::*;
+    use crate::signatures::ed25519;
+
+    impl Derivable for ed25519::SecretKey {
+        const ALLOW_NON_HARDENED: bool = false;
+        fn is_key_valid(key_bytes: &[u8; 33]) -> bool {
+            key_bytes[0] == 0
+        }
+        fn into_key(key_bytes: &[u8; 33]) -> Self {
+            debug_assert_eq!(0, key_bytes[0]);
+            let sk_bytes: &[u8; 32] = unsafe { &*(key_bytes[1..].as_ptr() as *const [u8; 32]) };
+            ed25519::SecretKey::from_bytes(sk_bytes)
+        }
+        fn add_key(_key_bytes: &mut [u8; 33], _parent_key: &[u8; 33]) -> bool {
+            true
+        }
+        fn calc_non_hardened_data(_key_bytes: &[u8; 33]) -> [u8; 33] {
+            unreachable!()
```

Comment:
If both the trait and the type are public, this actually is reachable? Maybe at least put a message in the panic?

Reply:
Maybe instead there should be a separate trait, not implemented for this type.

Reply:
Done.

Reply:
I don't know how to abstract it away with a trait. Need to think about it a bit more.

Reply:
You would create a new trait with that fn and impl it only for types that can perform that fn.

Reply:
But the whole point of the changes was to give a concrete type, so there is no runtime checking.